glean/lang/lsif/Data/LSIF/Angle.hs

```diff
@@ -0,0 +1,255 @@
+-- uses unique ids for facts. We can use that to simplify processing
+--
+-- vertexes correspond to facts holding new data
+-- edges relate existnig vertex facts to each other (with new facts)
+--
+-- Currently we emit one predicate per fact in LSIF. Which is verbose. We should
+-- group by predicate first, before generating the final value.
+--
+toAngle :: LSIF -> Value
+toAngle (LSIF facts) = Array $ V.concatMap V.fromList $
+  evalState (V.mapM factToAngle facts) empty
+
+-- | Process each key/fact pair
+factToAngle :: LSIF.KeyFact -> State Env [Value]
+factToAngle (KeyFact (Id id_) MetaData{..}) = do
+  setRoot projectRoot
+  predicate "lsif.Metadata.1" ([
+    "lsifVersion" .= version,
+    "projectRoot" .= string projectRoot, -- src.File
+    "positionEncoding" .= positionEncoding
+    ] ++ (case toolInfo of -- optional
+            Nothing -> []
+            Just ToolInfo{..} ->
+              ["toolInfo" .= object [
+                "toolName" .= toolName,
+                "toolArgs" .= toolArgs,
+                "version" .= version
+              ]]
+      ))
+
+factToAngle (KeyFact (Id id_) Project{..}) =
+  predicate "lsif.Project.1" [ "kind" .= fromEnum kind ]
+
+factToAngle (KeyFact _ PackageInformation{..}) =
+  predicate "lsif.PackageInformation.1" [
+    "name" .= name,
+    "manager" .= manager,
+    "version" .= version
+  ]
+
+-- LSIF Documents are uri/language pairs.
+factToAngle (KeyFact n Document{..}) = do
+  path <- filterRoot uri
+  predicate "lsif.Document.1"
+    [ "file" .= object [
+            factId n, -- generates a new src.File fact with this id
+            "key" .= path
+         ],
+      "language" .= fromEnum language
+    ]
+
+-- Push document or project identifier onto open stack. This is list of
+-- documents or projects for which facts may still be emitted.
+factToAngle (KeyFact _ (Event Begin scope n)) = [] <$ case scope of
+  ProjectScope -> insertProject n
+  DocumentScope -> insertDocument n -- should be src.File id of previous $event
+
+-- We don't currently use restult sets to group things. Ignore
+factToAngle (KeyFact _ ResultSet) = pure []
+
+-- Definition ranges. They include the identifier , span of the identifier and
+-- full range of the def
+factToAngle (KeyFact n (SymbolRange range Definition{..})) =
+  predicateId "lsif.Range.1" n
+    [ "range" .= toRange range, -- lsif.RangeSpan
+      "fullRange" .= toRange fullRange,
+      "text" .= string tagText
+    ]
+
+-- Same as for definitions, but for decls
+factToAngle (KeyFact n (SymbolRange range Declaration{..})) =
+  predicateId "lsif.Range.1" n
+    [ "range" .= toRange range,
+      "fullRange" .= toRange fullRange,
+      "text" .= string tagText
+    ]
+
+-- This is a reference identifier, just the span of something that is an xref
+-- n.b. we don't know what it refers to yet (hence no fullRange)
+factToAngle (KeyFact n (SymbolRange range Reference{..})) =
+  predicateId "lsif.Range.1" n
```

Comment:
as above, shouldn't this be a different type? And how do we get to the target symbol?